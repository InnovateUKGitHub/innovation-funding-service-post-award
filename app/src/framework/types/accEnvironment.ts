import { AccEnvironment } from "@framework/constants/enums";

/**
 * Parse an incoming ACC Environment.
 *
 * @param value The ACC Environment to parse.
 * @returns The parsed log level value. Defaults to BASE if not found.
 */
export function parseAccEnvironment(value?: string | AccEnvironment): AccEnvironment {
  switch ((value || "").toLowerCase()) {
    case AccEnvironment.BASE:
      return AccEnvironment.BASE;
    case AccEnvironment.AT:
      return AccEnvironment.AT;
    case AccEnvironment.CUSTOM:
      return AccEnvironment.CUSTOM;
    case AccEnvironment.DEMO:
      return AccEnvironment.DEMO;
    case AccEnvironment.DEV:
      return AccEnvironment.DEV;
    case AccEnvironment.PERF:
      return AccEnvironment.PERF;
    case AccEnvironment.PREPROD:
      return AccEnvironment.PREPROD;
    case AccEnvironment.PROD:
      return AccEnvironment.PROD;
    case AccEnvironment.SYSINT:
      return AccEnvironment.SYSINT;
    case AccEnvironment.UAT:
      return AccEnvironment.UAT;
    case AccEnvironment.LOCAL:
      return AccEnvironment.LOCAL;
    case AccEnvironment.CAPDEV:
      return AccEnvironment.CAPDEV;
    default:
      return AccEnvironment.UNKNOWN;
  }
}
