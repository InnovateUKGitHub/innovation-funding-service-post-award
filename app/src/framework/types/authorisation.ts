import { ProjectRole } from "@framework/constants";
import { IRoleInfo } from "@server/features/projects/getAllProjectRolesForUser";

type AvailableAuthRoles = "Fc" | "Mo" | "Pm" | "PmOrMo" | "PmAndFc" | "Unknown" | "SuperAdmin";

export function getAuthRoles(role: ProjectRole): Record<`is${AvailableAuthRoles}`, boolean> {
  // Note: As 'Unknown' there is never an overlap so we check against itself
  const isUnknown = (role & ProjectRole.Unknown) === ProjectRole.Unknown && role === ProjectRole.Unknown;

  const isFc = !!(role & ProjectRole.FinancialContact);
  const isPm = !!(role & ProjectRole.ProjectManager);
  const isMo = !!(role & ProjectRole.MonitoringOfficer);

  const isSuperAdmin = isFc && isPm && isMo;

  const isPmOrMo = isPm || isMo;
  const isPmAndFc = !isMo && isFc && isPm;

  return {
    isUnknown,
    isMo,
    isFc,
    isPm,
    isSuperAdmin,
    isPmAndFc,
    isPmOrMo,
  };
}

class RoleChecker {
  private readonly permissions: ProjectRole;
  constructor(roles: ProjectRole) {
    this.permissions = roles;
  }

  public hasRole(role: ProjectRole): boolean {
    return this.hasAllRoles(role);
  }

  public hasOnlyRole(role: ProjectRole): boolean {
    return this.permissions === role;
  }

  public hasAllRoles(...roles: ProjectRole[]): boolean {
    const combinedRoles = this.combineRoles(roles);
    return (this.permissions & combinedRoles) === combinedRoles;
  }

  public hasAnyRoles(...roles: ProjectRole[]): boolean {
    const combinedRoles = this.combineRoles(roles);
    return (this.permissions & combinedRoles) !== ProjectRole.Unknown;
  }

  public getRoles() {
    return this.permissions;
  }

  private combineRoles(roles: ProjectRole[]) {
    return roles.reduce((combined, r) => (combined |= r), ProjectRole.Unknown);
  }
}

// This class acts as a helper for the role information cached on the server
export class Authorisation {
  constructor(roles: { [key: string]: IRoleInfo }) {
    this.permissions = roles;
  }

  public readonly permissions: { [key: string]: IRoleInfo };

  public forPartner(projectId: string, partnerId: string) {
    const roles = this.getRolesForPartner(projectId, partnerId);
    return new RoleChecker(roles);
  }

  public forProject(projectId: string) {
    const roles = this.getRolesForProject(projectId);
    return new RoleChecker(roles);
  }

  private getRolesForPartner(projectId: string, partnerId: string) {
    const project = this.permissions && this.permissions[projectId];
    if (project) {
      const partner = project.partnerRoles[partnerId];
      if (partner !== undefined) {
        return partner;
      }
    }
    return ProjectRole.Unknown;
  }

  private getRolesForProject(projectId: string) {
    const project = this.permissions && this.permissions[projectId];
    if (project) {
      return project.projectRoles;
    }
    return ProjectRole.Unknown;
  }
}
