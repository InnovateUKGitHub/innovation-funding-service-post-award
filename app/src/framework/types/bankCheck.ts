//
// Request/response definitions for the SIL bank checks
//

export interface BankDetails {
  readonly sortcode: string;
  readonly accountNumber: string;
}

export interface BankCheckCondition {
  readonly severity: "error" | "warning" | "information";
  readonly code: number;
  readonly description: string;
}

export interface BankCheckResult {
  readonly conditions: BankCheckCondition | BankCheckCondition[] | undefined;
}

export interface BankCheckValidationResult extends BankCheckResult {
  readonly checkPassed: boolean;
  readonly iban: string | null;
}

export interface BankCheckAddress {
  readonly organisation: string;
  readonly buildingName: string;
  readonly street: string;
  readonly locality: string;
  readonly town: string;
  readonly postcode: string;
}

export interface AccountDetails {
  readonly sortcode: string;
  readonly accountNumber: string;
  readonly companyName: string;
  readonly registrationNumber: string;
  readonly firstName: string;
  readonly lastName: string;
  readonly address: BankCheckAddress;
}

type Score = 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | null;
type MatchFlag = "Match" | "No Match" | null;

export interface BankCheckVerificationResult extends BankCheckResult {
  readonly addressScore: Score;
  readonly companyNameScore: Score;
  readonly personalDetailsScore: Score;
  readonly regNumberScore: MatchFlag;
}
