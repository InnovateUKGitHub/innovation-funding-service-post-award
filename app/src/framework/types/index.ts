export * from "./authorisation";
export * from "../dtos";
export * from "../constants";
export * from "./IAppError";
export * from "./IContext";
export * from "./IContext";
export * from "./ILinkInfo";
export * from "./IPicklistEntry";
export * from "./IUser";
export * from "./permisionGroupIndentifier";
