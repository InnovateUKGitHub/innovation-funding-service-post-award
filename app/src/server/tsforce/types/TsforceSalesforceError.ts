interface TsforceSalesforceError {
  fields: string[];
  message: string;
  statusCode: string;
}

export { TsforceSalesforceError };
