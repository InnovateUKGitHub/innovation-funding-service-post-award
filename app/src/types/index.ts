export * from "./authorisation";
export * from "./dtos";
export * from "./constants";
export * from "./IUser";
export * from "./IAppError";
