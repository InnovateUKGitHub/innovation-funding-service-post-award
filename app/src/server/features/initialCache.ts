import { PermissionGroupIdentifier } from "@framework/constants/enums";
import { IContext } from "@framework/types/IContext";
import { GetClaimStatusesQuery } from "@server/features/claims/getClaimStatusesQuery";
import { GetAllRecordTypesQuery } from "@server/features/general/getAllRecordTypesQuery";
import { GetPermissionGroupQuery } from "@server/features/general/getPermissionGroupsQuery";
import { GetMonitoringReportStatusesQuery } from "@server/features/monitoringReports/getMonitoringReportStatusesQuery";
import { GetPcrParticipantSizesQuery } from "@server/features/pcrs/getPcrParticipantSizesQuery";
import { GetPcrPartnerTypesQuery } from "@server/features/pcrs/getPcrPartnerTypesQuery";
import { GetPcrProjectLocationsQuery } from "@server/features/pcrs/getPcrProjectLocationsQuery";
import { GetPcrProjectRolesQuery } from "@server/features/pcrs/getPcrProjectRolesQuery";
import { GetPcrSpendProfileCapitalUsageTypesQuery } from "@server/features/pcrs/getPcrSpendProfileCapitalUsageTypesQuery";
import { GetPcrSpendProfileOverheadRateOptionsQuery } from "@server/features/pcrs/getPcrSpendProfileOverheadRateOptionsQuery";
import { GetPcrStatusesQuery } from "@server/features/pcrs/getPcrStatusesQuery";
import { GetUnfilteredCostCategoriesQuery } from "./claims/getCostCategoriesQuery";
import { QueryBase } from "./common/queryBase";

type ICacheQuery = [string, QueryBase<unknown>];

/**
 * @description This describes the name of the cache and query it is caching, this will get executed on app starting.
 *
 * Queries must be non user specific! The idea here is to isolate all initial queries rather than clutter the server
 */
const initialCacheQueries: ICacheQuery[] = [
  ["Permission group cache", new GetPermissionGroupQuery(PermissionGroupIdentifier.ClaimsTeam)],
  ["Record type cache", new GetAllRecordTypesQuery()],
  ["Cost Categories cache", new GetUnfilteredCostCategoriesQuery()],
  ["Claim Statuses cache", new GetClaimStatusesQuery()],
  ["Monitoring Report Status cache", new GetMonitoringReportStatusesQuery()],
  ["PCR - Status cache", new GetPcrStatusesQuery()],
  ["PCR - Project Roles cache", new GetPcrProjectRolesQuery()],
  ["PCR - Partner Types cache", new GetPcrPartnerTypesQuery()],
  ["PCR - Participant Sizes cache", new GetPcrParticipantSizesQuery()],
  ["PCR - Project Locations cache", new GetPcrProjectLocationsQuery()],
  ["PCR - Spend Profile Capital Usage Types cache", new GetPcrSpendProfileCapitalUsageTypesQuery()],
  ["PCR - Spend Profile Overhead Rate Options cache", new GetPcrSpendProfileOverheadRateOptionsQuery()],
];

export const fetchCaches = async ({ config, logger, runQuery }: IContext) => {
  const isHTTPS = config.webserver.url.startsWith("https://");

  // Note: We only cache on PROD (when live), otherwise we have a slower dev experience
  if (!isHTTPS) return;

  await initialCacheQueries.forEach(async ([cacheKey, cacheQuery]) => {
    try {
      logger.info(`Priming cache '${cacheKey}'`);

      await runQuery(cacheQuery);

      logger.info(`Successfully primed cache '${cacheKey}'`);
    } catch (error) {
      logger.error(`Unable to prime cache '${cacheKey}'`, error);
    }
  });
};
