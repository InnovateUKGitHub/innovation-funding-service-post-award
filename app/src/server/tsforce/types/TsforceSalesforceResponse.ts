import { TsforceSalesforceError } from "./TsforceSalesforceError";

type TsforceSalesforceResponse = TsforceSuccessfulSalesforceResponse | TsforceUnsuccessfulSalesforceResponse;

interface TsforceSuccessfulSalesforceResponse {
  success: true;
  id: string;
  body: AnyObject;
  queryParameters: AnyObject;
  url: string;
}
interface TsforceUnsuccessfulSalesforceResponse {
  success: false;
  errors: TsforceSalesforceError[];
}

export { TsforceSalesforceResponse };
