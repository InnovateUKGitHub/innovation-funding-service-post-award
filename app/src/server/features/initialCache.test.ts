import { IContext } from "@framework/types";
import { fetchCaches } from "./initialCache";

const logger = {
  info: jest.fn(),
  error: jest.fn(),
};

const config = {
  serverUrl: "https://ukri.org",
};

const runQuery = jest.fn();

const context = {
  logger,
  config,
  runQuery,
} as unknown as IContext;

beforeEach(jest.clearAllMocks);

describe("initialCache", () => {
  it("should initialise the cache", async () => {
    await fetchCaches(context);
    expect(runQuery.mock.calls).toMatchSnapshot();
    expect(logger.info.mock.calls).toMatchSnapshot();
  });

  it("should initialise log errors if the attempt to query fails", async () => {
   const failedContext = {
    ...context,
    runQuery: jest.fn().mockImplementation(() => {
throw new Error("bang!");
})
   } as unknown as IContext;
    await fetchCaches(failedContext);

    expect(logger.error.mock.calls).toMatchSnapshot();
  });

  it("should not initialise the cache if not prod (not https domain)", async () => {
    const notProdContext = { ...context, config: { serverUrl: "http://acc-dev.ukri.org" } } as unknown as IContext;
    await fetchCaches(notProdContext);
    expect(runQuery).not.toBeCalled();
    expect(logger.info).not.toBeCalled();
  });
});
