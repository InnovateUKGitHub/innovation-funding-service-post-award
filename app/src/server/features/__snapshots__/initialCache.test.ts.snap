// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`initialCache should initialise log errors if the attempt to query fails 1`] = `
[
  [
    "Unable to prime cache 'Permission group cache'",
    [Error: bang!],
  ],
  [
    "Unable to prime cache 'Record type cache'",
    [Error: bang!],
  ],
  [
    "Unable to prime cache 'Cost Categories cache'",
    [Error: bang!],
  ],
  [
    "Unable to prime cache 'Claim Statuses cache'",
    [Error: bang!],
  ],
  [
    "Unable to prime cache 'Monitoring Report Status cache'",
    [Error: bang!],
  ],
  [
    "Unable to prime cache 'PCR - Status cache'",
    [Error: bang!],
  ],
  [
    "Unable to prime cache 'PCR - Project Roles cache'",
    [Error: bang!],
  ],
  [
    "Unable to prime cache 'PCR - Partner Types cache'",
    [Error: bang!],
  ],
  [
    "Unable to prime cache 'PCR - Participant Sizes cache'",
    [Error: bang!],
  ],
  [
    "Unable to prime cache 'PCR - Project Locations cache'",
    [Error: bang!],
  ],
  [
    "Unable to prime cache 'PCR - Spend Profile Capital Usage Types cache'",
    [Error: bang!],
  ],
  [
    "Unable to prime cache 'PCR - Spend Profile Overhead Rate Options cache'",
    [Error: bang!],
  ],
]
`;

exports[`initialCache should initialise the cache 1`] = `
[
  [
    GetPermissionGroupQuery {
      "identifier": 1,
    },
  ],
  [
    GetAllRecordTypesQuery {},
  ],
  [
    GetUnfilteredCostCategoriesQuery {},
  ],
  [
    GetClaimStatusesQuery {
      "key": "ClaimStatus",
    },
  ],
  [
    GetMonitoringReportStatusesQuery {
      "key": "MonitoringReports",
    },
  ],
  [
    GetPcrStatusesQuery {
      "key": "PCRStatuses",
    },
  ],
  [
    GetPcrProjectRolesQuery {
      "key": "PCRProjectRoles",
    },
  ],
  [
    GetPcrPartnerTypesQuery {
      "key": "PCRPartnerTypes",
    },
  ],
  [
    GetPcrParticipantSizesQuery {
      "key": "PCRParticipantSize",
    },
  ],
  [
    GetPcrProjectLocationsQuery {
      "key": "PCRProjectLocations",
    },
  ],
  [
    GetPcrSpendProfileCapitalUsageTypesQuery {
      "key": "PCRSpendProfileCapitalUsageTypes",
    },
  ],
  [
    GetPcrSpendProfileOverheadRateOptionsQuery {
      "key": "PCRSpendProfileOverheadRateOptions",
    },
  ],
]
`;

exports[`initialCache should initialise the cache 2`] = `
[
  [
    "Priming cache 'Permission group cache'",
  ],
  [
    "Priming cache 'Record type cache'",
  ],
  [
    "Priming cache 'Cost Categories cache'",
  ],
  [
    "Priming cache 'Claim Statuses cache'",
  ],
  [
    "Priming cache 'Monitoring Report Status cache'",
  ],
  [
    "Priming cache 'PCR - Status cache'",
  ],
  [
    "Priming cache 'PCR - Project Roles cache'",
  ],
  [
    "Priming cache 'PCR - Partner Types cache'",
  ],
  [
    "Priming cache 'PCR - Participant Sizes cache'",
  ],
  [
    "Priming cache 'PCR - Project Locations cache'",
  ],
  [
    "Priming cache 'PCR - Spend Profile Capital Usage Types cache'",
  ],
  [
    "Priming cache 'PCR - Spend Profile Overhead Rate Options cache'",
  ],
  [
    "Successfully primed cache 'Permission group cache'",
  ],
  [
    "Successfully primed cache 'Record type cache'",
  ],
  [
    "Successfully primed cache 'Cost Categories cache'",
  ],
  [
    "Successfully primed cache 'Claim Statuses cache'",
  ],
  [
    "Successfully primed cache 'Monitoring Report Status cache'",
  ],
  [
    "Successfully primed cache 'PCR - Status cache'",
  ],
  [
    "Successfully primed cache 'PCR - Project Roles cache'",
  ],
  [
    "Successfully primed cache 'PCR - Partner Types cache'",
  ],
  [
    "Successfully primed cache 'PCR - Participant Sizes cache'",
  ],
  [
    "Successfully primed cache 'PCR - Project Locations cache'",
  ],
  [
    "Successfully primed cache 'PCR - Spend Profile Capital Usage Types cache'",
  ],
  [
    "Successfully primed cache 'PCR - Spend Profile Overhead Rate Options cache'",
  ],
]
`;
